File: juno/indicators/stoch.py
from collections import deque
from decimal import Decimal

from .sma import Sma


# Full Stochastic Oscillator
class Stoch:
    k: Decimal = Decimal('0.0')
    d: Decimal = Decimal('0.0')

    _k_high_window: deque[Decimal]
    _k_low_window: deque[Decimal]

    _k_sma: Sma
    _d_sma: Sma

    _t: int = 0
    _t1: int
    _t2: int
    _t3: int

    def __init__(self, k_period: int, k_sma_period: int, d_sma_period: int) -> None:
        if k_period < 1:
            raise ValueError(f'Invalid period ({k_period})')

        self._k_high_window = deque(maxlen=k_period)
        self._k_low_window = deque(maxlen=k_period)

        self._k_sma = Sma(k_sma_period)
        self._d_sma = Sma(d_sma_period)

        self._t1 = k_period
        self._t2 = self._t1 + k_sma_period - 1
        self._t3 = self._t2 + d_sma_period - 1

    @property
    def maturity(self) -> int:
        return self._t3

    @property
    def mature(self) -> bool:
        return self._t >= self._t3

    def update(self, high: Decimal, low: Decimal, close: Decimal) -> tuple[Decimal, Decimal]:
        self._t = min(self._t + 1, self._t3)

        self._k_high_window.append(high)
        self._k_low_window.append(low)

        if self._t >= self._t1:
            max_high = max(self._k_high_window)
            min_low = min(self._k_low_window)
            fast_k = 100 * (close - min_low) / (max_high - min_low)

            self._k_sma.update(fast_k)

            if self._t >= self._t2:
                self._d_sma.update(self._k_sma.value)

            if self._t >= self._t3:
                self.k = self._k_sma.value
                self.d = self._d_sma.value

        return self.k, self.d

File: juno_rs/src/indicators/stoch.rs
use super::{
    sma::{Sma, SmaParams},
    MA,
};
use bounded_vec_deque::BoundedVecDeque;
use std::cmp::min;

pub struct Stoch {
    pub k: f64,
    pub d: f64,
    k_high_window: BoundedVecDeque<f64>,
    k_low_window: BoundedVecDeque<f64>,
    k_sma: Sma,
    d_sma: Sma,
    t: u32,
    t1: u32,
    t2: u32,
    t3: u32,
}

impl Stoch {
    pub fn new(k_period: u32, k_sma_period: u32, d_sma_period: u32) -> Self {
        let t1 = k_period;
        let t2 = t1 + k_sma_period - 1;
        let t3 = t2 + d_sma_period - 1;
        Self {
            k: 0.0,
            d: 0.0,
            k_high_window: BoundedVecDeque::new(k_period as usize),
            k_low_window: BoundedVecDeque::new(k_period as usize),
            k_sma: Sma::new(&SmaParams {
                period: k_sma_period,
            }),
            d_sma: Sma::new(&SmaParams {
                period: d_sma_period,
            }),
            t: 0,
            t1,
            t2,
            t3,
        }
    }

    pub fn maturity(&self) -> u32 {
        self.t3
    }

    pub fn mature(&self) -> bool {
        self.t >= self.t3
    }

    pub fn update(&mut self, high: f64, low: f64, close: f64) {
        self.t = min(self.t + 1, self.t3);

        self.k_high_window.push_back(high);
        self.k_low_window.push_back(low);

        if self.t >= self.t1 {
            let max_high = self.k_high_window.iter().cloned().fold(f64::MIN, f64::max);
            let min_low = self.k_low_window.iter().cloned().fold(f64::MAX, f64::min);
            let fast_k = 100.0 * (close - min_low) / (max_high - min_low);

            self.k_sma.update(fast_k);

            if self.t >= self.t2 {
                self.d_sma.update(self.k_sma.value);
            }

            if self.t >= self.t3 {
                self.k = self.k_sma.value;
                self.d = self.d_sma.value;
            }
        }
    }
}
